cpp/inc/bond/ext/grpc/unary_call.h

```diff
@@ -20,31 +20,70 @@
 #include <bond/ext/grpc/detail/io_manager_tag.h>
 
 #include <boost/assert.hpp>
+#include <boost/optional.hpp>
 #include <atomic>
+#include <memory>
 
 namespace bond { namespace ext { namespace gRPC {
 
 namespace detail {
 
+    template <typename TRequest, typename TResponse, typename TThreadPool>
+    struct service_unary_call_data;
+
     /// @brief Implementation class that holds the state associated with a
     /// single async, unary call.
     ///
-    /// It manages its own lifetime: to send a response, it enques itself in
-    /// the completion queue. When that sending of the response is done and
-    /// it is dequeued from the completion queue, it deletes itself.
+    /// Two different object participate in shared ownership of instances of
+    /// this class: the user-facing unary_call object and the
+    /// response_sent_tag helper struct.
+    ///
+    /// To ensure that it stays alive while sending a response,
+    /// unary_call_impl creates a response_sent_tag with a shared_ptr to
+    /// itself that it enques in the completion queue. When that sending of
+    /// the response is done and it is dequeued from the completion queue,
+    /// response_sent_tag clears its shared_ptr to the unary_call_impl.
+    ///
+    /// The user-facing unary_call object also has a shared_ptr to an
+    /// instance of unary_call_impl to keep the state alive while the call
+    /// object is itself alive.
     template <typename TRequest, typename TResponse>
-    struct unary_call_impl final : detail::io_manager_tag
+        struct unary_call_impl final : std::enable_shared_from_this<unary_call_impl<TRequest, TResponse>>
     {
+        /// @brief Handles keeping a unary_call_impl instance alive while a
+        /// response it being sent.
+        struct response_sent_tag final : detail::io_manager_tag
```

Comment:
I added this class to attempt to separate the `unary_call_impl` data from the response sending behavior. Let me know whether or not you think this makes things clearer or muddier.